### DISCUSSION TITLE
User auth redirect

### DISCUSSION THREAD

**Discussion by birdpump** [pocketbase/pocketbase#690] [Q&A] [answered]
Hello, I am making a site with pocketbase, express, and nodejs, and I am confused on how to best implement a feature where if you access the main index page and are not authenticated previously, you should be redirected to a login page. I made the login page that uses the pocketbase js sdk to authenticate the user by email fully in the browser, and doesn't send any requests the the server. I think the token needs to be sent to the server with the request and the server will redirect if not present. But I have tried setting a cookie during login that the server checks, and this works, but I don't think it is the right, or secure solution. Thanks for the help

**Comment by ganigeorgiev** [comment, marked as answer]
After authentication, `client.authStore` will use the previously issued token and will add a `Authorizaiton: User [TOKEN]` header to every following request (if not disabled explicitly).

Using a cookie is OK as long as you have `SameSite=strict` and `Secure` flags enabled.

The default JS SDK Authstore comes with helper methods to generate this cookie:

```js
// update the store with the parsed data from the cookie string
client.authStore.loadFromCookie('pb_auth=...');

// exports the store data as cookie, with option to extend the default SameSite, Secure, HttpOnly, Path and Expires attributes
client.authStore.exportToCookie({ httpOnly: false }); // Output: 'pb_auth=...'
```

How this cookie will be stored and loaded depends where the PocketBase client is initialized.

1. If it is only on the server, then use the express request/response objects.
2. If it is only on the browser, then use document.cookie.
3. If you need the PocketBase client to be initialized both in the browser and server, then you could combine 1&2 (the cookie's HttpOnly flag need to be disabled)

The SSR integration is similar to your use case. For examples, you check - https://github.com/pocketbase/js-sdk#ssr-integration.

For a mixed browser and server approach, you could do something similar as shown in https://github.com/pocketbase/js-sdk/issues/29#issuecomment-1229959528.

**Reply by birdpump** [reply]
Thank you very much. I have a better understanding now and ill give it a try.

**Reply by birdpump** [reply]
Sorry, one more thing. Do the above commands take the token in local storage and write it as a cookie? Because I am trying those now, but I don't think I am doing something right because its not saving anything. Does the authStore need to be defined?

**Reply by ganigeorgiev** [reply]
The default AuthStore doesn't send cookies. The above 2 methods are just helpers to parse an external cookie and export the current auth state as cookie string. 

> Because I am trying those now, but I don't think I am doing something right because its not saving anything.

I'm confused. Where and what do you expect to be saved? You first have to specify how do you want to interact with the PocketBase API:

1. browser client -> PocketBase API
2. node server -> PocketBase API
3. both 1) and 2)

The easiest way to have route permissions checks in the node server is to use 2). This means that the client-side will interact only with the node server and the node server will decide whether to call the PocketBase API or not to check the access. This is similar to the SSR integration https://github.com/pocketbase/js-sdk#ssr-integration.

If you want  a mixed approach then you'll have to manage 2 different JS SDK client instances - 1 that is working with document.cookie in the browser and a second one intialized on the node server that is using the express request/response headers for reading and sending the cookies (similar to https://github.com/pocketbase/js-sdk/issues/29#issuecomment-1229959528)

**Reply by birdpump** [reply]
I basically just want to set the auth token as a cookie in the client browser after a user logs in. Then on the server check if it is valid during a request for a page.

**Reply by ganigeorgiev** [reply]
@marchyeler 

It seems that you want the mixed approach. Without a code sample, I'm not sure how to help, but here is some guidance (_I haven't tested it_):

1. Client-side:

```js
import PocketBase from 'pocketbase';

// client setup
const client = new PocketBase("http://127.0.0.1:8090");

client.authStore.loadFromCookie(document.cookie);

client.authStore.onChange(() => {
    document.cookie = client.authStore.exportToCookie({ httpOnly: false });
})
// ---


// example login
await client.users.authViaEmail("test@example.com", "123456");
```

2. Server-side:

```js
import express from "express";
import PocketBase from "pocketbase";

const app = express();

// middleware to register the PocketBase client
app.use(async (req, res, next) => {
    // attach the PocketBase client to the request object
    req.pbClient = new PocketBase("http://127.0.0.1:8090");

    // load the store data from the request cookie string
    req.pbClient.authStore.loadFromCookie(req.headers.cookie || '');
    
    // on AuthStore change update the response cookie header
    req.pbClient.authStore.onChange(() => {
        res.setHeader("Set-Cookie", req.pbClient.authStore.exportToCookie({ httpOnly: false }));
    });

    try {
        // get an up-to-date auth state by refreshing the loaded auth model (if any)
        if (req.pbClient.authStore.isValid) {
            await req.pbClient.users.refresh();
        }
    } catch (_) {
        // clear the auth store on failed refresh
        req.pbClient.authStore.clear();
    }

    next();
})

app.get("/", async function (req, res) {
    if (!req.pbClient.authStore.isValid) {
        return res.sendStatus(401); // unauthorized
    }

    res.send("Hello World!");
});

app.listen(3000, function () {
    console.log("App started...");
});
```

**Reply by birdpump** [reply]
Okay, perfect. It works now with the code above. The problem was that it was saving the token in local storage, but it wasn't making the cookie. Also just to be sure, this line on the server side checks if the token is valid from pocketbase correct? `req.pbClient.authStore.isValid`

**Reply by ganigeorgiev** [reply]
> Also just to be sure, this line on the server side checks if the token is valid from pocketbase correct? req.pbClient.authStore.isValid

I've updated the above example to clear the store in case the refresh() call fails.You can trust the check because we are verifying the token with the refresh() call in the middleware (if the token is invalid or expired refresh will fail).

**Reply by birdpump** [reply]
@ganigeorgiev hello again, after beginning work on my app again after an update, im having an issue again. I narrowed it down to this 
`try {
        if (pb.authStore.isValid) {
            await pb.collection('users').authRefresh();` the authRefresh is making the token not valid for some reason. Thanks for the help

**Reply by ganigeorgiev** [reply]
@marchyeler Could you elaborate a little more on what do you mean by "_the authRefresh is making the token not valid for some reason_"? I've tried the following simplified example and it works as expected:

```js
const pb = new PocketBase("http://127.0.0.1:8090");

const user = await pb.collection("users").authWithPassword("test@example.com", "1234567890");

console.log("BEFORE REFRESH:", pb.authStore.isValid, pb.authStore.token);

const result = await pb.collection("users").authRefresh()

console.log("AFTER REFRESH:", pb.authStore.isValid, pb.authStore.token);
```
(in both cases isValid is true, just the token could be different after the refresh call)

Make sure that your call doesn't end up in the `catch` block (aka. try adding a console.log/debug point inside it):

```js
try {
    // get an up-to-date auth state by refreshing the loaded auth model (if any)
    if (req.pbClient.authStore.isValid) {
        await req.pbClient.users.refresh();
    }
} catch (err) {
    console.log("REFRESH ERROR:" err);

    // clear the auth store on failed refresh
    req.pbClient.authStore.clear();
}
```

Also note that PocketBase v0.8 works only with PocketBase JS SDK v0.8+.

**Reply by ganigeorgiev** [reply]
I've also noticed that in your example you are using `pb.collection(...).authRefresh()` instead of `req.pb.collection(...).authRefresh()`, or in other words when PocketBase SDK is initialized in the node server make sure that you are creating a new instance for each request and not using a global shared variable.

**Reply by birdpump** [reply]
here is my full code, also the pb.authStore.onChange is giving me some errors 
```
const express = require('express');
const app = express();
const http = require('http');
const server = http.createServer(app);
const { Server } = require("socket.io");

const io = new Server(server);
app.use(express.static('public'))
const bodyParser = require('body-parser');
app.use(bodyParser.urlencoded({ extended: true }));

//cookie prase
const cookieParser = require("cookie-parser");


//pocketbase config
const PocketBase = require('pocketbase/cjs');
const pb = new PocketBase('http://127.0.0.1:8090');



app.use(async (req, res, next) => {

    // load cookie 
    pb.authStore.loadFromCookie(req.headers.cookie, key = 'pb_auth')
    console.log(req.headers.cookie);

     pb.authStore.onChange(() => {
         res.setHeader("Set-Cookie", pb.authStore.exportToCookie({ httpOnly: false }, key = 'pb_auth'));
    })
    
    console.log(pb.authStore.isValid);
    try {
        //  auth state by refreshing the loaded auth model
        if (pb.authStore.isValid) {
            await pb.collection('users').authRefresh();
        }
    } catch (_) {
        // clear auth store
        pb.authStore.clear();
    }

    next();
})

app.get("/", async function (req, res) {
    return res.sendFile(__dirname + '/src/index.html');

});

app.get("/app", async function (req, res) {
    if (!pb.authStore.isValid) {
        console.log(pb.authStore.isValid);
        return res.redirect('http://localhost:3000/');
    }
    console.log(pb.authStore.isValid);
    res.sendFile(__dirname + '/src/app.html');
});


app.get("/login", async function (req, res) {
    if (!pb.authStore.isValid) {
        return res.sendFile(__dirname + '/src/login.html');
    }

    res.redirect('http://localhost:3000/app');
});


app.get("/register", async function (req, res) {
    return res.sendFile(__dirname + '/src/register.html');

});


server.listen(3000, () => {
    console.log('listening on *:3000');
});
```

**Reply by ganigeorgiev** [reply]
@marchyeler The above is incorrect. You shouldn't use a global PocketBase instance on the node server and should attach it to the `req` object as shown in the initial example - https://github.com/pocketbase/pocketbase/discussions/690#discussioncomment-3782980.

The reason why is wrong to have a single PocketBase client instance inside the node server is that you'll always have a single AuthStore state and requests from different users will overwrite each others (aka. you have many users -> one server; in the browser its ok to have a global instance because whatever is happening in your browser it is isolated and will not affect other users).

**Reply by birdpump** [reply]
okay I see, its as if its a single user

**Reply by ganigeorgiev** [reply]
> okay I see, its as if its a single user

Yes, exactly.

The above code could be changed to something like this:

```js
const express = require('express');
const app = express();
const http = require('http');
const server = http.createServer(app);
const { Server } = require("socket.io");

const io = new Server(server);
app.use(express.static('public'))
const bodyParser = require('body-parser');
app.use(bodyParser.urlencoded({ extended: true }));

//cookie prase
const cookieParser = require("cookie-parser");

//pocketbase config
const PocketBase = require('pocketbase/cjs');

app.use(async (req, res, next) => {
    req.pb = new PocketBase('http://127.0.0.1:8090');

    // load cookie
    req.pb.authStore.loadFromCookie(req.headers.cookie)
    console.log(req.headers.cookie);

     req.pb.authStore.onChange(() => {
         res.setHeader("Set-Cookie", req.pb.authStore.exportToCookie({ httpOnly: false }));
    })

    console.log(req.pb.authStore.isValid);
    try {
        //  auth state by refreshing the loaded auth model
        if (req.pb.authStore.isValid) {
            await req.pb.collection('users').authRefresh();
        }
    } catch (err) {
        console.log("REFRESH error (mostly likely due to expired token)", err);

        // clear auth store
        req.pb.authStore.clear();
    }

    next();
})

app.get("/", async function (req, res) {
    return res.sendFile(__dirname + '/src/index.html');

});

app.get("/app", async function (req, res) {
    if (!req.pb.authStore.isValid) {
        console.log(req.pb.authStore.isValid);
        return res.redirect('http://localhost:3000/');
    }
    console.log(req.pb.authStore.isValid);
    res.sendFile(__dirname + '/src/app.html');
});

app.get("/login", async function (req, res) {
    if (!req.pb.authStore.isValid) {
        return res.sendFile(__dirname + '/src/login.html');
    }

    res.redirect('http://localhost:3000/app');
});

app.get("/register", async function (req, res) {
    return res.sendFile(__dirname + '/src/register.html');

});

server.listen(3000, () => {
    console.log('listening on *:3000');
});

```

**Reply by birdpump** [reply]
okay that's odd, even with the above code it only works if a comment out the refresh part, it is causing some error 
```
const express = require('express');
const app = express();
const http = require('http');
const server = http.createServer(app);
const { Server } = require("socket.io");

const io = new Server(server);
app.use(express.static('public'))
const bodyParser = require('body-parser');
app.use(bodyParser.urlencoded({ extended: true }));

//cookie prase
const cookieParser = require("cookie-parser");

//pocketbase config
const PocketBase = require('pocketbase/cjs');

app.use(async (req, res, next) => {
    req.pb = new PocketBase('http://127.0.0.1:8090');

    // load cookie
    req.pb.authStore.loadFromCookie(req.headers.cookie, key = 'pb_auth')
    console.log(req.headers.cookie);

     req.pb.authStore.onChange(() => {
         res.setHeader("Set-Cookie", req.pb.authStore.exportToCookie({ httpOnly: false }, key = 'pb_auth'));
    })

    console.log(req.pb.authStore.isValid);
    // try {
    //     //  auth state by refreshing the loaded auth model
    //     if (req.pb.authStore.isValid) {
    //         await req.pb.collection('users').authRefresh();
    //     }
    // } catch (err) {
    //     console.log("REFRESH error (mostly likely due to expired token)", err);

    //     // clear auth store
    //     req.pb.authStore.clear();
    // }

    next();
})

app.get("/", async function (req, res) {
    return res.sendFile(__dirname + '/src/index.html');

});

app.get("/app", async function (req, res) {
    if (!req.pb.authStore.isValid) {
        console.log(req.pb.authStore.isValid);
        return res.redirect('http://localhost:3000/');
    }
    console.log(req.pb.authStore.isValid);
    res.sendFile(__dirname + '/src/app.html');
});

app.get("/login", async function (req, res) {
    if (!req.pb.authStore.isValid) {
        return res.sendFile(__dirname + '/src/login.html');
    }

    res.redirect('http://localhost:3000/app');
});

app.get("/register", async function (req, res) {
    return res.sendFile(__dirname + '/src/register.html');

});

server.listen(3000, () => {
    console.log('listening on *:3000');
});

```

**Reply by birdpump** [reply]
the issue that im getting is that it is clearing the token from the client when you try to access the /app page

**Reply by ganigeorgiev** [reply]
@marchyeler I think the reason is because of the `key = "pb_auth"` parameter you are setting when loading and exporting the cookie because I guess it results to `undefined` (if you want to pass a custom key you can just pass the string).

Try changing:

```
req.pb.authStore.loadFromCookie(req.headers.cookie, key = 'pb_auth')                                -> req.pb.authStore.loadFromCookie(req.headers.cookie)
res.setHeader("Set-Cookie", req.pb.authStore.exportToCookie({ httpOnly: false }, key = 'pb_auth')); -> res.setHeader("Set-Cookie", req.pb.authStore.exportToCookie({ httpOnly: false }));
```

If the above doesn't solve the issue then please post the refresh call error.

**Reply by ganigeorgiev** [reply]
Actually, the above will evaluate to the assigned value, not to `undefined`, but it is incorrect/strange way to pass arguments anyway:

```
function test (a, b) {
    console.log(a, b);
}

test("abc", foo = 123); // strangely works and is identical to test("abc", 123)
// output: "abc", 123
```

**Reply by birdpump** [reply]
still same error, that's what I get:
```
REFRESH error (mostly likely due to expired token) TypeError: req.pb.collection is not a function
    at C:\dev\#posse-chat\index.js:33:26
    at Layer.handle [as handle_request] (C:\dev\#posse-chat\node_modules\express\lib\router\layer.js:95:5)
    at trim_prefix (C:\dev\#posse-chat\node_modules\express\lib\router\index.js:328:13)
    at C:\dev\#posse-chat\node_modules\express\lib\router\index.js:286:9
    at Function.process_params (C:\dev\#posse-chat\node_modules\express\lib\router\index.js:346:12)
    at next (C:\dev\#posse-chat\node_modules\express\lib\router\index.js:280:10)
    at urlencodedParser (C:\dev\#posse-chat\node_modules\body-parser\lib\types\urlencoded.js:91:7)
    at Layer.handle [as handle_request] (C:\dev\#posse-chat\node_modules\express\lib\router\layer.js:95:5)
    at trim_prefix (C:\dev\#posse-chat\node_modules\express\lib\router\index.js:328:13)
    at C:\dev\#posse-chat\node_modules\express\lib\router\index.js:286:9
```

**Reply by ganigeorgiev** [reply]
Hm, it seems that you are using an older version of the JS SDK. `pb.collection()` is available only with PocketBase JS SDK v0.8.0+ (try running `npm update` and/or check your `package.json`).

**Reply by birdpump** [reply]
yeah im on "pocketbase": "^0.7.4"

**Reply by birdpump** [reply]
okay, that was it, it works now after updating, thanks for the help Gani

**Comment by birdpump** [comment]
still the same after updating though

**Reply by ganigeorgiev** [reply]
I cannot reproduce it. Make sure that you are using PocketBase v0.8.0 both for the backend and the SDK.

**Reply by ganigeorgiev** [reply]
Also please note that if you've previously used PocketBase v0.7.x you need to perform some manual migration steps - https://github.com/pocketbase/pocketbase/blob/master/CHANGELOG.md#migrate-from-v07x

**Reply by birdpump** [reply]
I have already migrated the login and everything, just got stuck on this
